### PR TITLE
GUI localization tweaks

### PR DESCRIPTION
**Pull request by elpaso**
Part of QEP https://github.com/qgis/QGIS-Enhancement-Proposals/issues/210

Locale support for numeric input and display: revision and enhancements

Grant proposal 2021

Affected widgets/classes:

    - QgisApp
    - QgsDecorationGridDialog
    - QgsHandleBadLayers
    - QgsPointRotationItem
    - QgsSnappingLayerTreeModel
    - QgsStatisticalSummaryDockWidget
    - QgsDirectoryParamWidget
    - QgsClassificationStandardDeviation
    - QgsLayoutRuler
    - QgsMeshRendererScalarSettingsWidget
    - QgsMeshRendererVectorSettingsWidget
    - QgsMeshVariableStrokeWidthWidget
    - QgsVectorLayerSaveAsDialog
    - QgsProcessingNumberParameterDefinitionWidget
    - QgsProcessingRangeParameterDefinitionWidget
    - QgsProcessingDistanceParameterDefinitionWidget
    - QgsAbstractRelationEditorWidget
    - QgsCoordinateOperationWidget
    - QgsDoubleValidator (added ::setMaxDecimals( int decimals ) )
    - QgsExtentWidget
    - QgsMapToolIdentify
    - QgsPropertyAssistantWidget
    - QgsRasterHistogramWidget
    - QgsRendererRasterPropertiesWidget
    - QgsDashSpaceWidget
    - QgsDataDefinedSizeLegendWidget
    - QgsGraduatedSymbolRendererWidget
    - QgsGraduatedSymbolRendererWidget

### CODE REVIEW

**Comment by elpaso** [comment]
@vcloarec I had to make a lot of adjustments to the meshes widgets, mind having a look please?

**Comment by elpaso** [comment]
CC @SrNetoChan

File: src/gui/mesh/qgsmeshrendererscalarsettingswidget.cpp [hunk 1]
```diff
@@ -44,10 +49,10 @@ QgsMeshRendererScalarSettingsWidget::QgsMeshRendererScalarSettingsWidget( QWidge
 
   // connect
   connect( mScalarRecalculateMinMaxButton, &QPushButton::clicked, this, &QgsMeshRendererScalarSettingsWidget::recalculateMinMaxButtonClicked );
-  connect( mScalarMinLineEdit, &QLineEdit::textChanged, this, &QgsMeshRendererScalarSettingsWidget::minMaxChanged );
-  connect( mScalarMaxLineEdit, &QLineEdit::textChanged, this, &QgsMeshRendererScalarSettingsWidget::minMaxChanged );
-  connect( mScalarMinLineEdit, &QLineEdit::textEdited, this, &QgsMeshRendererScalarSettingsWidget::minMaxEdited );
-  connect( mScalarMaxLineEdit, &QLineEdit::textEdited, this, &QgsMeshRendererScalarSettingsWidget::minMaxEdited );
+  connect( mScalarMinSpinBox, qOverload<double>( &QgsDoubleSpinBox::valueChanged ), this, [ = ]( double ) { minMaxChanged(); } );
+  connect( mScalarMaxSpinBox, qOverload<double>( &QgsDoubleSpinBox::valueChanged ), this, [ = ]( double ) { minMaxChanged(); } );
+  connect( mScalarMinSpinBox, &QgsDoubleSpinBox::editingFinished, this, &QgsMeshRendererScalarSettingsWidget::minMaxEdited );
```
Comment (vcloarec) [inline src/gui/mesh/qgsmeshrendererscalarsettingswidget.cpp:54]: this breaks the update of the classification when user change min/max

Comment (vcloarec) [reply src/gui/mesh/qgsmeshrendererscalarsettingswidget.cpp:54]: Before, when user change the value, even if he doesn't finished to enter the value, the classification was updated. I just see that `minMaxEdited` was always called.
I wonder was it is the best :  
* the classification is updated as before (if user type "12", first update for "1" then for "2")
* or wait for the spin box lost the focus or the user press enter.
I don't have a strong opinion on this.
For both case, it seems that  `minMaxChanged` does not serve a lot... Further more, it breaks your implementation: first update min/max without updating classification, then when `minMaxEdited` is called, as the min/max are the same, the classification is not updated.
I would just use `minMaxEdited` for both signals (or only for `editingFinished` if we don't  want update of the classification when editing)

Comment (elpaso) [reply src/gui/mesh/qgsmeshrendererscalarsettingswidget.cpp]: How about this: https://github.com/qgis/QGIS/pull/43831/commits/1600e25b47f6b30a46959e97e2ee7a9eed493e99 ?

Comment (vcloarec) [reply src/gui/mesh/qgsmeshrendererscalarsettingswidget.cpp]: I tried it and it seems ok, thanks !

Comment (elpaso) [reply src/gui/mesh/qgsmeshrendererscalarsettingswidget.cpp]: @vcloarec thank you for the review!

File: python/gui/auto_generated/qgsdoublevalidator.sip.in [hunk 2]
```diff
@@ -50,26 +50,38 @@ Constructor for QgsDoubleValidator.
 :param parent: parent object
 %End
 
-    QgsDoubleValidator( double bottom, double top, int decimal, QObject *parent );
+    QgsDoubleValidator( double bottom, double top, int decimals, QObject *parent );
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsdoublevalidator.sip.in:53]: This is an api break -- we can't change argument names without potentially breaking PyQGIS code which uses named arguments

File: python/gui/auto_generated/qgsdoublevalidator.sip.in [hunk 3]
```diff
@@ -50,26 +50,38 @@ Constructor for QgsDoubleValidator.
...
 .. versionadded:: 3.16
 %End
 
+    void setMaxDecimals( int maxDecimals );
+%Docstring
+Sets the number of decimals accepted by the validator to ``maxDecimals``.
+
+.. warning::
+
+   setting decimals overrides any custom regular expression that was previosly set
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsdoublevalidator.sip.in:79]:
```suggestion
   setting decimals overrides any custom regular expression that was previously set
```

File: src/ui/mesh/qgsmeshrenderervectorsettingswidgetbase.ui [hunk 4]
```diff
@@ -639,26 +706,16 @@
   <tabstop>mLineWidthSpinBox</tabstop>
   <tabstop>mColoringMethodComboBox</tabstop>
   <tabstop>mColorWidget</tabstop>
-  <tabstop>mColorRampShaderMinimumEditLine</tabstop>
-  <tabstop>mColorRampShaderMaximumEditLine</tabstop>
   <tabstop>mColorRampShaderLoadButton</tabstop>
-  <tabstop>mMinMagLineEdit</tabstop>
-  <tabstop>mMaxMagLineEdit</tabstop>
   <tabstop>mDisplayVectorsOnGridGroupBox</tabstop>
```
Comment (DelazJ) [inline src/ui/mesh/qgsmeshrenderervectorsettingswidgetbase.ui:710]: @elpaso, can you please add the new widgets and fix the tabstop in the .ui files? thanks.

Comment (elpaso) [reply src/ui/mesh/qgsmeshrenderervectorsettingswidgetbase.ui:710]: @DelazJ sorry but I don't know what you are talking about :)

Comment (DelazJ) [reply src/ui/mesh/qgsmeshrenderervectorsettingswidgetbase.ui:710]: This tabstops section lists the widgets in the order they will be accessed by users using their keyboard (the TAB key) to navigate through the GUI. So missing items are put to the end, hence an unordered and confusing navigation.
To fix it, open the ui in Qt Designer, Edit menu --> Edit tab order: you probably see that numbers next to the widgets are not in the correct order. click or right-click and you get access to ways to fix.

Comment (elpaso) [reply src/ui/mesh/qgsmeshrenderervectorsettingswidgetbase.ui:710]: @DelazJ thanks, I know what tabstops are, I just didn't recall changing them, the order probably broke when I replaced the widgets.

Comment (DelazJ) [reply src/ui/mesh/qgsmeshrenderervectorsettingswidgetbase.ui:710]: @elpaso Sorry for the explanation, then. It's unfortunately very frequent that when changes are done to ui, people do not fix that. and yes, Designer is the culprit as it places by default new items at the end regardless their real placement in the GUI.

Comment (elpaso) [reply src/ui/mesh/qgsmeshrenderervectorsettingswidgetbase.ui]: @DelazJ fixed, thank you for checking this PR.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**